lightning-invoice/src/payment.rs

```diff
@@ -0,0 +1,556 @@
+//! [`InvoicePayer`] so that such changes take effect before retrying.
+
+use crate::Invoice;
+
+use bitcoin::hashes::Hash;
+use bitcoin::hashes::sha256::Hash as Sha256;
+use bitcoin::secp256k1::key::PublicKey;
+
+use lightning::ln::{PaymentHash, PaymentSecret};
+use lightning::ln::channelmanager::{ChannelDetails, PaymentSendFailure};
+use lightning::ln::features::InvoiceFeatures;
+use lightning::ln::msgs::LightningError;
+use lightning::routing::router::{Route, RouteHint};
+use lightning::util::events::{Event, EventHandler};
+use lightning::util::logger::Logger;
+
+use std::collections::hash_map::{self, HashMap};
+use std::ops::Deref;
+use std::sync::Mutex;
+
+/// A utility for paying [`Invoice]`s.
+pub struct InvoicePayer<P: Deref, R, L: Deref, E>
+where
+	P::Target: Payer,
+	R: Router,
+	L::Target: Logger,
+	E: EventHandler,
+{
+	payer: P,
+	router: R,
+	logger: L,
+	event_handler: E,
+	invoice_cache: Mutex<HashMap<PaymentHash, (Invoice, usize)>>,
+	retry_attempts: usize,
+}
+
+/// A trait defining behavior of an [`Invoice`] payer.
+pub trait Payer {
+	/// Returns the payer's node id.
+	fn node_id(&self) -> PublicKey;
+
+	/// Returns the payer's channels.
+	fn first_hops(&self) -> Vec<ChannelDetails>;
+
+	/// Sends a payment over the Lightning Network using the given [`Route`].
+	fn send_payment(
+		&self, route: &Route, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>
+	) -> Result<(), PaymentSendFailure>;
+}
+
+/// A trait defining behavior for routing an [`Invoice`] payment.
+pub trait Router {
+	/// Finds a [`Route`] between `payer` and `payee` using the provided first and last hops for a
+	/// payment with the given values.
+	fn find_route(
+		&self,
+		payer: &PublicKey,
+		payee: &PublicKey,
+		payee_features: Option<InvoiceFeatures>,
+		first_hops: Option<&[&ChannelDetails]>,
+		last_hops: &[&RouteHint],
+		final_value_msat: u64,
+		final_cltv: u32,
+	) -> Result<Route, LightningError>;
+}
+
+/// An error that may occur when making a payment.
+#[derive(Debug)]
+pub enum PaymentError {
+	/// An error resulting from the provided [`Invoice`] or payment hash.
+	Invoice(&'static str),
+	/// An error occurring when finding a route.
+	Routing(LightningError),
+	/// An error occurring when sending a payment.
+	Sending(PaymentSendFailure),
+}
+
+impl<P: Deref, R, L: Deref, E> InvoicePayer<P, R, L, E>
+where
+	P::Target: Payer,
+	R: Router,
+	L::Target: Logger,
+	E: EventHandler,
+{
+	/// Creates an invoice payer.
+	pub fn new(payer: P, router: R, logger: L, event_handler: E) -> Self {
+		Self {
+			payer,
+			router,
+			logger,
+			event_handler,
+			invoice_cache: Mutex::new(HashMap::new()),
+			retry_attempts: 0,
+		}
+	}
+
+	/// Sets the maximum number of payment retry attempts before forwarding the
+	/// [`Event::PaymentPathFailed`] event to be handled by the decorated [`EventHandler`].
+	pub fn with_retry_attempts(self, retry_attempts: usize) -> Self {
+		Self { retry_attempts, ..self }
```

Comment:
what's the benefit of this method as opposed to taking `&mut self` and changing the `retry_attempts` property?

Reply:
By essentially making the field const, it limits the possible states the object can be in. For example, if a payment is retried less than the number of retry attempts, and then the latter is decreased, the code could potentially continue to retry the payment indefinitely (without adjusting the code accordingly).

It also makes for easier to read initialization. The object can be initialized in one statement and wouldn't need to be `mut`.

Reply:
Does it? It would allow a user to change the retry attempt count on the fly while payments are in-flight, it seems.

More generally, however, this is super annoying to map in bindings - many languages (JS/Java/etc) don't have any concept of move semantics/self-by-ownership, meaning there's ~no way to map this without clone-ing the `Payer`, which doesn't really make any logical sense. If we want it to be explicit, why not just put it in the constructor, otherwise I think we should make it take &mut, sadly.